controllers/topology/reconcile_state.go

```diff
@@ -61,6 +70,93 @@ func (r *ClusterReconciler) reconcileState(ctx context.Context, s *scope.Scope)
 	return r.reconcileMachineDeployments(ctx, s)
 }
 
+// Reconcile the Cluster shim, a temporary object used a mean to collect objects/templates
+// that might be orphaned in case of errors during the remaining part of the reconcile process.
+func (r *ClusterReconciler) reconcileClusterShim(ctx context.Context, s *scope.Scope) error {
+	shim := clusterShim(s.Current.Cluster)
+
+	// If we are going to create the InfrastructureCluster or the ControlPlane object, then
+	// add a temporary cluster-shim object and use it as an additional owner.
+	// This will ensure the objects will be garbage collected in case of errors in between
+	// creating InfrastructureCluster/ControlPlane objects and updating the Cluster with the
+	// references to above objects.
+	if s.Current.InfrastructureCluster == nil || s.Current.ControlPlane.Object == nil {
+		if err := r.Client.Create(ctx, shim); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "failed to create the cluster shim object")
+			}
+			if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shim), shim); err != nil {
+				return errors.Wrapf(err, "failed to read the cluster shim object")
+			}
+		}
+		// Enforce type meta back given that it gets blanked out by Get.
+		shim.Kind = "Secret"
+		shim.APIVersion = corev1.SchemeGroupVersion.String()
+
+		// Add the shim as a temporary owner for the InfrastructureCluster.
+		ownerRefs := s.Desired.InfrastructureCluster.GetOwnerReferences()
+		ownerRefs = append(ownerRefs, *ownerReferenceTo(shim))
+		s.Desired.InfrastructureCluster.SetOwnerReferences(ownerRefs)
+
+		// Add the shim as a temporary owner for the ControlPlane.
+		ownerRefs = s.Desired.ControlPlane.Object.GetOwnerReferences()
+		ownerRefs = append(ownerRefs, *ownerReferenceTo(shim))
+		s.Desired.ControlPlane.Object.SetOwnerReferences(ownerRefs)
+	}
+
+	// If the InfrastructureCluster and the ControlPlane objects has been already created
+	// in previous reconciliation, check if they have already been reconciled by the ClusterController
+	// by verifying the ownerReference for the Cluster is present
+	// In this case, if the ownerReference from the shim object is still there, we can
+	// now get rid of the shim object, garbage collecting all the orphaned objects eventually
+	// created in case of errors in between creating InfrastructureCluster/ControlPlane objects
+	// and updating the cluster with the references to above objects.
+	// If instead the ownerReference from the shim object has been already removed, it means that
+	// the shim objects has been already removed in previous reconciliations.
+	if s.Current.InfrastructureCluster != nil && s.Current.ControlPlane.Object != nil {
+		objectsReconciled := hasOwnerReferenceFrom(s.Current.InfrastructureCluster, s.Current.Cluster) &&
+			hasOwnerReferenceFrom(s.Current.ControlPlane.Object, s.Current.Cluster)
+		objectsStillOwnedByTheShim := hasOwnerReferenceFrom(s.Current.InfrastructureCluster, shim) ||
```

Comment:
```suggestion
		shimOwnsAtLeastOne := hasOwnerReferenceFrom(s.Current.InfrastructureCluster, shim) ||
```

Not sure about this check in general, we're saying if the Shim still owns either Infrastructure or ControlPlane  here, but shouldn't the shim own both before removing the shim altogether?

Reply:
Yes, it should own both, both given that this is used for cleaning up temporary things after the cluster owns both the objects, I changed && into || so we are cleaning up no matter of what the shim actually owns at this stage